### DISCUSSION TITLE
How to export and import the best model?

### DISCUSSION THREAD

**Discussion by MarioADguez** [winedarksea/AutoTS#73] [Q&A]
I writing to them because I need your help.

When I export the file pikel, the object is TemplateEvalObject and I need use the best model, how use TemplateEvalObject? I don't know. Or how use the method "import_results"? because I need the object AutoTS but my model return "Uninitiated AutoTS object".
in other words, I don't know how to import and export the best model to use it another time and not to train it continuously.

**Comment by winedarksea** [comment]
Does this help:
https://github.com/winedarksea/AutoTS/blob/master/extended_tutorial.md#deployment-and-template-importexport

**Reply by winedarksea** [reply]
Also checkout the new production_example in the main repo

**Reply by tmggr** [reply]
hello, 

I came across the same problem and i tried to follow the above examples and although i manage to use the model without retraining it, I can not find a way to update it. For instance, I want to perform 10 times One Step Ahead Forecasting after having identified a suitable model and basically bring it up to date each time without retraining all possible combinations. I can not find any way to do that without having to refit all models.
Is a  possible bypass to reinitialize an object but fix it to the selected model? If so can you give an example of how to do that?

Thank you in advance

**Reply by winedarksea** [reply]
@tmggr 
It sounds to me that what you should be doing is setting `forecast_length=10` and then just predicting once.

Or possibly what you are looking for is this:
https://github.com/winedarksea/AutoTS/blob/master/extended_tutorial.md#running-just-one-model
where you specify the model as the best_model from the AutoTS fit, then predict on a new df_train.

behind the scenes, some of the models are doing exactly what you are trying to do, forecasting 1 step ahead to produce that total length of 10 (Rolling Regression, WindowRegression, UnivariateRegression as well as models like ARIMA, all depending on parameters)

**Reply by tmggr** [reply]
my apologies, seems I was not clear on my question. Basically I would like to use it on a bench-marking setting. For that purpose I need the Rolling Cross Validation to be consistent on what is performed where. So i want to manually handle the testing of the best model, hence i keep an extra portion of the timeseries latest values. If i raise the forecast_length then if I have not made any mistakes prior, the predictions severely underperform, whether it is with one steap ahead evaluation or k-step ahead.

I followed the quoted instructions prior to my first post, that is why I sought your advice, cause it seems to me that the model does not update when specified directly. Even if I simply run the quoted example the results seems to be a small regression around the last value. Meaning, train set of first column has values ranging from 3.9 to 148.2, with last value at 22.0. If i copy paste and run your example code the 5 values range from 22.04 to 22.23. This unfortunately seems to be also the case for the rest of the columns.

So it seems to me that there is some kind of bug on either the training process compared to long forecast steps, or there is something missing from the example to update properly

**Reply by winedarksea** [reply]
It sounds like you are using the sample code with the "superfast" model list of naive-only models, which will of course choose naive models like you seem to be having. Try tuning the model_list and metric_weighting if you want different results. 

And it really isn't too surprising to me that models predicting a continual of the last value naive are performing well. From weather forecasting to gas prices to the stock market, last value naive forecasts (sometimes called "persistence" models) are often among the best performing models in the real world, and from cross validation that was likely confirmed here.

Could you provide some sample code? That might suggest if there is a bug.

**Comment by tmggr** [comment]
Good evening,

I reply to you via email to attach a ipnyb file of a snippet trying to set
up the process.
Apologies for the numerical example late at night after a long day I failed
to notice that the example was a naive model.
Along with the python script I attach some sample data I use to debug the
scripts. The first series I test and I'm not sure the models update or
predict properly is a randomly generated ARMA(1,1) model from statsmodel
functions, and a simple pmdarima autoARIMA performs better

Thank you VERY much for your time, and any insights would be extremely
valuable

**Reply by winedarksea** [reply]
How are you defining a "best" model here? How it looks in graphs? Or a specific metric?

In AutoTS, adjust the `metric_weighting` to match what you want will produce a model closer to what you are expecting (for example, spl to 0 if you don't care about upper/lower forecasts). Another important thing is adjusting the cross validation in AutoTS - it will choose a model based on the validation sections, and if those aren't representative, they won't deliver the best model.

Also, there's a fair amount of random chance when it comes to forecasting. "Perfect" results are impossible.

Finally, I didn't get any email from you. I am not sure how those are forwarded here.

**Comment by tmggr** [comment]
First time I used this method from here also :) but thankfully they are
forwarded!

As the "best" model in general I'm always choosing based on a metric
(usually RMSE, SMAPE or sMAPE). The graphs in the script were just for a
quick inspection.
In the script I attached I used on the "testing" set the model returned
from AutoTS in the best_model dictionary, while trying to build up a
processing pipeline to use on my main benchmarking dataset.

Thank you for your answers, I will go through your documentation again and
see what more adjustments I need to perform.
